### PR TITLE
Development

### PR DESCRIPTION
**Pull request by vrushbru**
Code refactor

### CODE REVIEW

File: api-spec.yaml [hunk 1]
```diff
@@ -15,41 +15,62 @@ servers:
...
+          content:
+            application/json:
+              schema:
+                type: object
+                properties:
+                  data:
+                    type: object
+                    properties:
+                      sessionId:
+                        description: Id of the quiz generated.
```
Comment (rdsarda) [inline api-spec.yaml:43]: Typo -- its the id of the session

File: api-spec.yaml [hunk 2]
```diff
@@ -70,8 +95,9 @@ paths:
               operationId: getQuiz
               parameters:
                 quizId: $response.body#/quizId
+            
         '400':
-          description: Bad Request
+          description: Bad Request
```
Comment (rdsarda) [inline api-spec.yaml:100]: extra space character at the end, can be removed

File: api-spec.yaml [hunk 3]
```diff
@@ -123,31 +151,76 @@ paths:
...
+             required:
+              - data
+             properties:
+              data:
+                type: object
+                required:
+                  - userTime
+                  - userAnswer
+                properties:
+                  userTime:
```
Comment (rdsarda) [inline api-spec.yaml:193]: please provide description. What is the unit of the time here?

File: api-spec.yaml [hunk 4]
```diff
@@ -123,31 +151,76 @@ paths:
...
+          schema:
+            type: integer
+        
+      requestBody:
+          description: Details of attempted question 
+          required: true
+          content:
+           application/json:
+            schema:
+             type: object
```
Comment (rdsarda) [inline api-spec.yaml:183]: Instead of inline definition of this request object, can we create a type and refer that here?

File: api-spec.yaml [hunk 5]
```diff
@@ -157,55 +230,37 @@ paths:
...
-                typeOfQuestion:
-                  type: string
       responses:
         '201':
           description: successful operation
           content:
             application/json:
               schema:
-                $ref: '#/components/schemas/Answer'
+                 $ref : '#/components/schemas/QuizStatus'
```
Comment (rdsarda) [inline api-spec.yaml:261]: unwanted extra space characters

File: api-spec.yaml [hunk 6]
```diff
@@ -217,30 +272,35 @@ paths:
...
             format: uuid
+        
       responses:
         '201':
           description: successful operation
           content:
             application/json:
               schema:
-                $ref: '#/components/schemas/Performance'
+                 $ref: '#/components/schemas/Performance'
```
Comment (rdsarda) [inline api-spec.yaml]: extra space in indentation

File: api-spec.yaml [hunk 7]
```diff
@@ -252,59 +312,69 @@ paths:
...
-          items:
-            $ref: '#/components/schemas/Multimedia'
-    Mode:
-      title: Mode
+          description: 'Name of the user'
+          example: 'Krishna Bharat'
+        avatarId:
+          type: string
+          description:  'Id of the avatar selected by user.'
+          example: 'f615d834-d8fd-11eb-b8bc-0242ac130003'
```
Comment (rdsarda) [inline api-spec.yaml:331]: are avatarIds uuids?

File: api-spec.yaml [hunk 8]
```diff
@@ -325,57 +395,126 @@ components:
...
+         type: array
+         items:
+           $ref: '#/components/schemas/Hint'
+​
+    Hint:
+      title: Hint
       type: object
       properties:
-        questionId:
+        id:
```
Comment (rdsarda) [inline api-spec.yaml:406]: what's the purpose of the id for an Hint?

File: api-spec.yaml [hunk 9]
```diff
@@ -325,57 +395,126 @@ components:
...
+      description: State of the question in the quiz.
+      enum:
+        - NOT_ATTEMPTED
+        - ATTEMPTED
+        - SKIPPED
+        
+  parameters:
+    SessionId:
+      in: header
+      name: sessionId
```
Comment (rdsarda) [inline api-spec.yaml:516]: Please rename it to `x-genkwiz-session-id`

File: api-spec.yaml [hunk 10]
```diff
@@ -325,57 +395,126 @@ components:
...
+        status:
+          $ref: '#/components/schemas/QuestionState'
+        correct:
+          type: boolean
+          description: 'true if user provided answer is correct.'
+        serialNumber:
+          type: integer
+          description: 'Serial number of the question in the Quiz.'
+          
+    Mode:
```
Comment (rdsarda) [inline api-spec.yaml:455]: What is the purpose of this object?

File: api-spec.yaml [hunk 11]
```diff
@@ -343,18 +481,40 @@ components:
...
+              
+    
+    QuestionState:
+      title: Question status
+      type: string
+      description: State of the question in the quiz.
+      enum:
+        - NOT_ATTEMPTED
+        - ATTEMPTED
+        - SKIPPED
```
Comment (rdsarda) [inline api-spec.yaml:511]: Please name the enums in all small case letters.

File: api-spec.yaml [hunk 12]
```diff
@@ -303,32 +398,75 @@ components:
...
+          description: 'true if user provided answer is correct.'
+        serialNumber:
+          type: integer
+          description: 'Serial number of the question in the Quiz.'
+          
+    Mode:
+      title: Mode
+      type: object
+      properties:
+        maxSore:
```
Comment (rdsarda) [inline api-spec.yaml:459]: typo - maxScore

File: api-spec.yaml [hunk 13]
```diff
@@ -343,18 +481,40 @@ components:
...
+            totalTime:
+              type: integer
+              format: int32
+            finalScore:
+              type: integer
+              format: int32
+              
+    
+    QuestionState:
+      title: Question status
```
Comment (rdsarda) [inline api-spec.yaml:505]: Please rename the `title` to `Question State`.

File: api-spec.yaml [hunk 14]
```diff
@@ -303,32 +398,75 @@ components:
...
+          description: 'Serial number of the question in the Quiz.'
+          
+    Mode:
+      title: Mode
+      type: object
+      properties:
+        maxSore:
+          type: integer
+          format: int32
+        maxTime:
```
Comment (rdsarda) [inline api-spec.yaml:462]: Is this timestamp or the amount of maximum time allowed to solve the question?

File: DB_Script.sql [hunk 15]
```diff
@@ -19,52 +19,47 @@ mode_id int not null,
...
 );
 
 create index index_question_hint on question_hint (question_id , hint_number);
 
-/*Purpose:Consists of all information about the multimedia available for individual questions.*/
+-- Purpose:Consists of all information about the multimedia available for individual questions.
 create table multimedia(
-id serial not null,
+id uuid not null,
 question_id int not null,
```
Comment (rdsarda) [inline DB_Script.sql:43]: What is the purpose of `question_id` here?

File: api-spec.yaml (hunk 10 above)
Comment (rdsarda) [inline api-spec.yaml:463]: What is the purpose of this object? I see that there is description, but how do we plan to use this?

File: api-spec.yaml [hunk 16]
```diff
@@ -427,16 +410,41 @@ components:
...
+      title: Question Status
       type: object
       properties:
         userAnswer:
           type: string
           description: User's question response.
         userTime:
           type: integer
-          description: 'Time taken by the user to answer the question.'
+          description: 'Time take by the user to answer this question in seconds.'
```
Comment (rdsarda) [inline api-spec.yaml:446]: typo `take` --> `taken`

File: api-spec.yaml [hunk 17]
```diff
@@ -427,16 +410,41 @@ components:
...
+      properties:
+        data:
+          type: object
+          required:
+            - userTime
+            - userAnswer
+          properties:
+            userTime:
+              type: integer
+              description: 'Time take by the user to answer this question in seconds'
```
Comment (rdsarda) [inline api-spec.yaml:427]: typo `take` --> `taken`

**Review by rdsarda — CHANGES_REQUESTED** [review]

File: DB_Script.sql (hunk 15 above)
Comment (nagnath-sawant) [reply DB_Script.sql:43]: As, multimedia is associated with question, we can keep this info for reference.

File: api-spec.yaml (hunk 10 above)
Comment (nagnath-sawant) [reply api-spec.yaml:463]: This is part of Question response, not sure about the earlier thoughts but few fields like Max Time and Max score can be utilized while displaying question.

File: DB_Script.sql (hunk 15 above)
Comment (nagnath-sawant) [reply DB_Script.sql]: Removed it.

File: api-spec.yaml (hunk 17 above)
Comment (nagnath-sawant) [reply api-spec.yaml]: done

File: api-spec.yaml (hunk 16 above)
Comment (nagnath-sawant) [reply api-spec.yaml]: done

File: api-spec.yaml (hunk 8 above)
Comment (nagnath-sawant) [reply api-spec.yaml]: Used in question response to get more details about quiz mode

**Review by rdsarda — APPROVED** [review]

**Pull request merged**